internal/sourceanalysis/rust.go

```diff
@@ -0,0 +1,275 @@
+package sourceanalysis
+
+import (
+	"bytes"
+	"debug/dwarf"
+	"debug/elf"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/osv-scanner/internal/cachedregexp"
+	"github.com/google/osv-scanner/internal/sourceanalysis/ar"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/reporter"
+	"github.com/ianlancetaylor/demangle"
+)
+
+const RustFlagsEnv = "RUSTFLAGS=-C opt-level=3 -C debuginfo=1"
+const RustLibExtension = ".rcgu.o/"
+
+func rustAnalysis(r reporter.Reporter, pkgs []models.PackageVulns, source models.SourceInfo) {
+	binaryPaths, err := rustBuildSource(r, source)
+	if err != nil {
+		r.PrintError(fmt.Sprintf("failed to build cargo/rust project from source: %s", err))
+		return
+	}
+
+	isCalledMap := map[string]bool{}
+
+	for _, path := range binaryPaths {
+		if strings.HasSuffix(path, ".rlib") {
+			// Is a library, so need an extra step to extract the object binary file before passing to parseDWARFData
+			objFilePath, err := extractRlibArchive(path)
+			if err != nil {
+				r.PrintError(fmt.Sprintf("failed to analyse '%s': %s", path, err))
+				continue
+			}
+			// TODO: Do we need to care about error on deletion here?
+			defer os.Remove(objFilePath)
+		}
+		calls, err := parseDWARFData(r, path)
+		if err != nil {
+			r.PrintError(fmt.Sprintf("failed to analyse '%s': %s", path, err))
+			continue
+		}
+
+		for _, pv := range pkgs {
+			for _, v := range pv.Vulnerabilities {
+				for _, a := range v.Affected {
+					// Example of RUSTSEC function level information:
+					//
+					// "affects": {
+					//     "os": [],
+					//     "functions": [
+					//         "smallvec::SmallVec::grow"
+					//     ],
+					//     "arch": []
+					// }
+					ecosystemAffects, ok := a.EcosystemSpecific["affects"].(map[string]interface{})
+					if !ok {
```

Comment:
Can we print an error in these cases? Also wonder if we can define these in a struct and ummarshal them from a map somehow.

Reply:
Not sure how an error would work here, some ecosystem specific data will either not exist (most e.g. GHSA advisories), or have a different format (GHSA also has different function info compared to RUSTSEC, but since GHSA just copies it from RUSTSEC, just supporting RUSTSEC format seems fine)

Reply:
was just thinking on if there are ways we can validate that we're unmarshalling things correctly with the correct types. e.g. detecting if a type (e.g. `functions`) has changed from array to something else unexpectedly. 

Since this is `ecosystem_specific`, this should be specific to and defined by Rust and be consistent. 

Not a big deal, just a minor defensive thing to be adding here. Could just be a TODO.